src/mqt/core/_core/__init__.pyi

```diff
@@ -1 +1,733 @@
+    def add_ancillary_register(self, nc: int, reg_name: str) -> None: ...
+    def append_operation(self, op: Operation) -> None: ...
+    def add_variable(self, var: Expression | float) -> None: ...
+    def add_variables(self, vars_: list[Expression | float]) -> None: ...
+    def instantiate(self, params: dict[Variable, float]) -> None: ...
+    def is_variable_free(self) -> bool: ...
+    def get_variables(self) -> list[Variable]: ...
+    def initialize_io_mapping(self) -> None: ...
+    @overload
+    def from_file(self, filename: str) -> None: ...
+    @overload
+    def from_file(self, filename: str, format_: str) -> None: ...
+    @overload
+    def dump(self) -> str: ...
+    @overload
+    def dump(self, filename: str, format_: str) -> None: ...
+    def to_open_qasm(self) -> str: ...
+    def __getitem__(self, key: int) -> Operation: ...
+
+class Control:
+    @overload
+    def __init__(self, qubit: int, /) -> None: ...
+    @overload
+    def __init__(self, qubit: int, control_type: ControlType, /) -> None: ...
+    @property
+    def control_type(self) -> ControlType: ...
+    @control_type.setter
+    def control_type(self, value: ControlType) -> None: ...
+    @property
+    def qubit(self) -> int: ...
+    @qubit.setter
+    def qubit(self, value: int) -> None: ...
+
+class ControlType:
+    __members__: ClassVar[dict[ControlType, int]] = ...  # read-only
+
+    Pos: ClassVar[ControlType] = ...
+    Neg: ClassVar[ControlType] = ...
+
+class Variable:
+    def __init__(self, name: str, /) -> None: ...
+    @property
+    def name(self) -> str: ...
+    def __eq__(self, arg: object) -> bool: ...
+    def __ne__(self, arg: object) -> bool: ...
+    def __lt__(self, arg: object) -> bool: ...
+    def __gt__(self, arg: object) -> bool: ...
+
+class Term:
+    @overload
+    def __init__(self, coeff: float, /) -> None: ...
+    @overload
+    def __init__(self, coeff: float, var: Variable, /) -> None: ...
+    @property
+    def coefficient(self) -> float: ...
+    @property
+    def variable(self) -> Variable: ...
+    def has_zero_coefficient(self) -> bool: ...
+    def add_coefficient(self, coeff: float) -> None: ...
+    def total_assignment(self, assignment: dict[Variable, int], /) -> float: ...
+    def evaluate(self, assignment: dict[Variable, int], /) -> float: ...
+    def __mul__(self, arg: float) -> Term: ...
+    def __rmul__(self, arg: float) -> Term: ...
+    def __true_div__(self, arg: float) -> Term: ...
+    def __rtrue_div__(self, arg: float) -> Term: ...
+
+class Expression:
+    @overload
+    def __init__(self) -> None: ...
+    @overload
+    def __init__(self, term: Term, /) -> None: ...
+    @overload
+    def __init__(self, coeff: float, /) -> None: ...
+    @overload
+    def __init__(self, expr: list[Term]) -> None: ...
+    @property
+    def terms(self) -> list[Term]: ...
+    @property
+    def constant(self) -> float: ...
+    @constant.setter
+    def constant(self, value: float) -> None: ...
```

Comment:
Seeing these kind of overloads here, I am really wondering whether it would be feasible to unify them to functions that accept, in this example, a `float` or a `Term`. But that might not be to easy to handle on the bindings side, right?
Maybe with lambdas and variants?
The above applies to many functions in throughout the bindings.

Reply:
I tried to stay close to the C++ implementation. Do you think much is gained by unifying them?

Reply:
Not too sure. I think as a dynamically typed language, Python just makes it easy to provide these unified functions and it seems a little cleaner to me. However, in the end, it doesn't really matter too much I guess.